gc_registry/certificate/services.py

```diff
@@ -304,40 +304,199 @@ def process_certificate_action(
+    if (
+        certificate_query.certificate_period_start
+        and certificate_query.certificate_period_end
+    ):
+        if (
+            certificate_query.certificate_period_start
+            > certificate_query.certificate_period_end
+        ):
+            logger.error(
+                "Certificate period start must be lower than certificate period end"
+            )
+            return False
+
+    # If using either bundle quantity or percentage, only one can be used
+    if (
+        certificate_query.certificate_quantity
+        and certificate_query.certificate_bundle_percentage
+    ):
+        logger.error(
+            "Only one of certificate quantity or percentage can be used for each query."
+        )
+        return False
+
+    # Bundle percentage must be between 0 and 100
+    if certificate_query.certificate_bundle_percentage:
+        if (
+            certificate_query.certificate_bundle_percentage <= 0
+            or certificate_query.certificate_bundle_percentage > 100
+        ):
+            logger.error("Certificate percentage must be between 0 and 100")
+            return False
+
+    return True
+
+
+def apply_bundle_quantity_or_percentage(
+    certificates_from_query: list[GranularCertificateBundle],
+    certificate_bundle_action: GranularCertificateActionBase,
+    write_session: Session,
+    read_session: Session,
+    esdb_client: EventStoreDBClient,
+) -> list[GranularCertificateBundle]:
+    """Apply the bundle quantity or percentage to the certificates from the query.
+
+    For each GC Bundle returned from the query, this function will split the bundle
+    if the desired GC quantity or percentage is less than the GC Bundle quantity, otherwise
+    it will return the GC Bundle unsplit.
+
+    Args:
+        certificates_from_query (list[GranularCertificateBundle]): The certificates from the query
+        certificate_bundle_action (GranularCertificateAction): The certificate action
+        write_session (Session): The database write session
+        read_session (Session): The database read session
+        esdb_client (EventStoreDBClient): The EventStoreDB client
+
+    Returns:
+        list[GranularCertificateBundle]: The list of certificates to transfer, split if required
+                                         such that the quantity of each bundle is equal to or
+                                         less than the desired bundle quantity, if provided, or
+                                         the percentage of the total certificates in the bundle.
+
+    """
+    # Just return the certificates from the query if no quantity or percentage is provided
+    if (certificate_bundle_action.certificate_quantity is None) & (
+        certificate_bundle_action.certificate_bundle_percentage is None
+    ):
+        return certificates_from_query
+
+    certificates_to_transfer = []
+
+    if certificate_bundle_action.certificate_quantity is not None:
```

Comment:
Because this might be operating on more than one GC bundle, if the `certificate_quantity` param is set then it's going to pull that number of certificates from each bundle that are selected from, this list is just a way of generalising whether they chose quantity or percentage. It will be clearer once I change the variable naming